rfd/0163-vnet.md

```diff
@@ -0,0 +1,811 @@
+us past the problem of acquiring root privileges and lets us focus on validating VNet as a tool.
+In the [Alternatives to osascript](#alternatives-to-osascript) section, we describe other approaches
+which offer better UX at the cost of development time.
+
+#### Guiding the user towards VNet
+
+Before VNet, the user would access web and TCP apps through the unified resources view in Connect:
+
+<img width="871" alt="Unified Resources view in Connect v15.0.0" src="assets/0163-apps-in-connect-v15.png">
+
+Clicking "Launch" next to a web app would launch this app in the browser through the `/web/launch`
+URL of the proxy service. The user could also set up a local proxy for the specific app by clicking
+on the three dots menu and choosing "Set up connection". For TCP apps, clicking "Connect" would set
+up a local proxy for that app.
+
+This changes with the introduction of VNet. Here's a screenshot from Teleport Connect in cluster
+`teleport-local.dev` running VNet:
+
+<img width="871" alt="Unified Resources view in Connect with VNet turned on" src="assets/0163-apps-with-vnet.png">
+
+Web apps now have two actions available in the three dots menu: "Connect and copy VNet address" and
+"Connect to local port". "Connect to local port" starts a regular local proxy. This is kept for
+users who might not be able to use VNet, either because they lack root privileges or VNet would
+conflict with other software running on the computer.
+
+"Connect and copy VNet address" opens the VNet panel and shows the root privileges prompt. While the
+prompt is shown, the VNet panel is in the "starting" state, explaining what VNet is.
+
+<img width="871" alt="VNet starting" src="assets/0163-vnet-starting.png">
+
+Once VNet starts, the VNet address of the app is copied to clipboard along with an in-app
+notification about the clipboard contents. Afterwards the button becomes just "Copy VNet address"
+when VNet is running. Copying the address copies `spec.public_addr` but with HTTP as the protocol.
+
+For TCP apps, the main "Connect" button is the equivalent of "Connect and copy VNet address". From
+the three dots menu, the user can select the sole option "Connect to local port". The address
+displayed on the card of a TCP app is the `spec.public_addr` of the app.
+
+Selecting an app from the search bar performs the same action as clicking on the main button in the
+app card.
+
+#### Managing VNet
+
+<img width="871" alt="VNet in Connections" src="assets/0163-vnet-in-connections.png">
+
+VNet has a persistent place in the UI in the connection list in top left. From there, the user can
+see VNet status and start or stop VNet, similar to how they can disconnect existing database, SSH
+and Kube connections.
+
+Clicking on the VNet item opens a dedicated VNet panel within the connection list. The panel
+displays the DNS zones that are currently being proxied. Any errors related to starting or stopping
```

Comment:
I'm not sure how feasible it'll be to show currently proxied DNS zones. The long-running privileged process could change them any time there's a new login. We don't want to talk with that process because of the overhead required.

I guess whenever the user logs in to a new cluster, Connect could simply refetch details about DNS zones for all active profiles and then update the list in the UI. This is a bit racy as the list fetched by the privileged process can in theory be different than the list fetched by the Electron app, but it should be good enough for now.

(BTW, tsh and Connect still use separate tsh home dirs, see https://github.com/gravitational/teleport/issues/25806)

Reply:
We could use the same code the get the list of DNS zones we *should* be proxying in both the normal and root processes and show this, assuming the root process is doing the right thing

Reply:
We'll respect `TELEPORT_HOME` to use the right directory, let's make sure it's propagated to the admin process correctly